### ISSUE TITLE
[Feature request] Mail notification on comment

### ISSUE THREAD

**Issue by 2FWAH** [PrivateBin/PrivateBin#307]
Hello,

It would be awesome to be able to receive a mail notification when a new comment is posted in a paste.

How this could be implemented?
I'm thinking about a (php?) script triggered when the comment is posted.
Recipient email addresses could be defined during the paste creation (obviously optional) and stored encrypted with the message.

Should the notification email contains the (full) link to the paste?
It's better for usability but some user would probably prefer to send the link over a more secure channel.
The full link (including the key) is know by the client, when posting the paste the client could send the full link to the server.
Obviously a malicious server will be able to store the key and decrypt the paste.

To summarize:
- Ability to enable/disable entire email notification system in the config file (public instances will probably want to disable this feature)
- When posting comment a checkbox "Send notification by email" should be present if email addresses were defined during the creation of the paste.
- A second checkbox: "Send link in the notification email. WARNING: The server will be able to see the key".

What do you think about this idea?

**Comment by rugk** [comment]
> Should the notification email contains the (full) link to the paste?

I'd say just no, don't do this. The user may just remember the name. Also mail transport encryption is still not as secure as simple website HTTPS, e.g.

But the general idea is nice. Maybe another point though: Should the mail be stored e2e encrypted in the paste?
* Pro: It does not need to be on plaintext on the server. When a comment is made, it may just have to be sent to the server again and it processes it and forgets it afterwards.
* Con: Anyone who opens the paste can extract the mail, as it is of course JS-encrypted.

**Comment by 2FWAH** [comment]
IMO to follow the zero-knowledge approach we should store the email addresses encrypted.

When posting a comment the client will trigger a script by a POST requests.
The request contains:
- ID of the paste (NOT the key)
- email addresses retrieved from the encrypted paste.
- Some token to prevent anyone to trigger the sending script?
This token could be stored both in clear in the sever (as a metadata of the paste or in a global "db") and encrypted in the past. Thus only clients with the right key will be able to trigger it.
An option should be added to disable this feature in case of public instance for example.

Looks good for you?

**Comment by 2FWAH** [comment]
It's true that sending emails is a privacy nightmare, but this feature would obviously be optional (maybe with a warning).

Example of typical use: having a conversation with an user over PrivateBin by successive comments. The issue is that users don't know when a new comment is posted, so users need to open the link often to check for new comment.

Instead of a mail notification system we can think about a hook system. Hooks will be triggered on event (paste creation, comment posted, etc) thus the server admin will be able to use any notification system.
Actually this can be done by watching the discussion directory of the paste with systemd path unit.

**Comment by elrido** [comment]
The hook system is in part what is planned in relation to the paste manager feature (hopefully 1.4).
